File: back/coupon/src/controllers/coupon.controller.ts
// coupon.controller.ts
import { Controller, Post, Body, Logger, HttpException, HttpStatus } from '@nestjs/common';
import { CouponService } from '../services/coupon.service';
import { UseCouponDto, CouponResponseDto } from '../dtos/coupon.dto';

@Controller('coupon')
export class CouponController {
 private readonly logger = new Logger(CouponController.name);

 constructor(private readonly couponService: CouponService) {}

 @Post('generate')
 async generateCoupon(): Promise<CouponResponseDto> {
   this.logger.log('쿠폰 생성 요청 받음');
   try {
     const couponCode = await this.couponService.generateAndValidateCoupon();
     this.logger.log(`생성된 쿠폰 코드: ${couponCode}`);
     return { 
       message: '쿠폰이 발급되었습니다.',
       coupon_code: couponCode 
     };
   } catch (error) {
     this.logger.error(`쿠폰 생성 실패: ${error.message}`);
     throw new HttpException(error.message, HttpStatus.BAD_REQUEST);
   }
 }

 @Post('use')
 async useCoupon(@Body() useCouponDto: UseCouponDto): Promise<CouponResponseDto> {
   try {
     return await this.couponService.useCoupon(useCouponDto.coupon_code);
   } catch (error) {
     throw new HttpException(error.message, HttpStatus.BAD_REQUEST);
   }
 }
}

File: Backend/src/main/java/com/example/backend/controller/CouponController.java
package com.example.backend.controller;

import com.example.backend.dto.CouponResponseDto;
import com.example.backend.dto.UseCouponDto;
import com.example.backend.service.CouponService;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;

@RestController
@RequestMapping("/coupon")
public class CouponController {

    private final CouponService couponService;

    public CouponController(CouponService couponService) {
        this.couponService = couponService;
    }

    @PostMapping("/generate")
    public ResponseEntity<CouponResponseDto> generateCoupon(@RequestParam String email) {
        try {
            String couponCode = couponService.generateAndAssignCoupon(email);
            return ResponseEntity.ok(new CouponResponseDto("쿠폰 발급 완료", couponCode));
        } catch (IllegalArgumentException e) {
            return ResponseEntity.badRequest().body(new CouponResponseDto(e.getMessage(), null));
        }
    }

    @PostMapping("/use")
    public ResponseEntity<CouponResponseDto> useCoupon(@RequestBody UseCouponDto dto) {
        try {
            String result = couponService.useCoupon(dto.getCouponCode());
            return ResponseEntity.ok(new CouponResponseDto(result, dto.getCouponCode()));
        } catch (IllegalArgumentException | IllegalStateException e) {
            return ResponseEntity.badRequest().body(new CouponResponseDto(e.getMessage(), dto.getCouponCode()));
        }
    }
}